AX: Optimize hot functions AXObjectCache::{get(Node&), getOrCreate(Node&)} by inlining get inside getOrCreate and eliminating one hashmap lookup

twilco opened the pull request:
#### fd8b56688bda9cfc9527916226d9854a756ef727
<pre>
AX: Optimize hot functions AXObjectCache::{get(Node&amp;), getOrCreate(Node&amp;)} by inlining get inside getOrCreate and eliminating one hashmap lookup
<a href="https://bugs.webkit.org/show_bug.cgi?id=298158">https://bugs.webkit.org/show_bug.cgi?id=298158</a>
<a href="https://rdar.apple.com/159526201">rdar://159526201</a>

Reviewed by Joshua Hoffman.

This commit implements three optimizations to make AXObjectCache::get and AXObjectCache::getOrCreate faster:

  1. Rename existing data structure AXObjectCache::m_nodeObjectMapping to AXObjectCache::m_nodeIdMapping, and make a
     new data structure with the name AXObjectCache::m_nodeObjectMapping that points from a Node&amp; directly to an Ref&lt;AccessibilityObject&gt;.
     This removes one HashMap lookup in AXObjectCache::get. m_renderObjectMapping and m_widgetObjectMapping were renamed
     to m_renderObjectIdMapping and m_widgetIdMapping to stay consistent with m_nodeIdMapping. The cost of this new
     data structure is very small on most sites, and even in the most extreme case of html.spec.whatwg.org, only adds
     0.56mb extra memory usage. I believe this is worth the cost for what we get (results below).

  2. Inline the `get` part of `getOrCreate(Node&amp;)`, and move the non-inlined part to new method `getOrCreateSlow`.
     Based on 60 seconds of using VoiceOver on html.spec.whatwg.org, we hit this new inlined fast path 98.9% of the
     time (~51,114,000 fast path hits vs. `590,000` slow-path hits). This required creation of a new method called
     exportedGetOrCreate which is outlined for consumers outside WebCore, as inlined functions cannot be exported.

  3. Add a new `AXObjectCache::getOrCreate(Element&amp;)` method, allowing us to skip an `is&lt;Document&gt;` check in contexts
     where we&apos;ve already casted a `Node` into an `Element`.

This eliminates 1532 samples (1.5 seconds) of main-thread time based on 60 seconds of using VoiceOver on html.spec.whatwg.org.

Places that used `getOrCreate` are now changed to import `AXObjectCacheInlines.h` instead of `AXObjectCache.h`.

* Source/WebCore/accessibility/AXListHelpers.cpp:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::cacheAndInitializeWrapper):
(WebCore::AXObjectCache::exportedGetOrCreate):
(WebCore::AXObjectCache::getOrCreateSlow):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::onPaint const):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXObjectCacheInlines.h:
(WebCore::AXObjectCache::getOrCreate):
* Source/WebCore/accessibility/AXSearchManager.cpp:
* Source/WebCore/accessibility/AccessibilityImageMapLink.cpp:
* Source/WebCore/accessibility/AccessibilityListBoxOption.cpp:
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
* Source/WebCore/accessibility/AccessibilityMenuListOption.cpp:
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::nextSibling const):
* Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
* Source/WebCore/inspector/InspectorOverlay.cpp:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/PrintContext.cpp:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::axObjectForElement const):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::getAccessibilityObjectForNode):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(-[WKPDFPluginAccessibilityObject parent]):
(-[WKPDFPluginAccessibilityObject accessibilityFocusedUIElement]):
(-[WKPDFPluginAccessibilityObject accessibilityAssociatedControlForAnnotation:]):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::accessibilityCoreObject):
* Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
(-[WKAccessibilityPDFDocumentObject accessibilityFocusedUIElement]):
(-[WKAccessibilityPDFDocumentObject accessibilityParent]):
(-[WKAccessibilityPDFDocumentObject accessibilityAssociatedControlForAnnotation:]):

Canonical link: <a href="https://commits.webkit.org/299439@main">https://commits.webkit.org/299439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/32fc2b3a779a89abdc9b90b43b46cbeb8535705c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118769 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38450 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29101 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124954 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70830 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120647 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39146 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47032 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90141 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59651 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121722 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31191 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106482 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70647 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30251 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24595 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68613 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100634 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24783 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128008 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45676 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34480 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98788 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46041 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102702 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98570 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25105 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44011 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22017 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42207 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45546 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51224 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45010 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48356 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46696 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua reviewed (approved):
This is a great perf boost!

twilco committed:
AX: Optimize hot functions AXObjectCache::{get(Node&), getOrCreate(Node&)} by inlining get inside getOrCreate and eliminating one hashmap lookup
https://bugs.webkit.org/show_bug.cgi?id=298158
rdar://159526201

Reviewed by Joshua Hoffman.

This commit implements three optimizations to make AXObjectCache::get and AXObjectCache::getOrCreate faster:

  1. Rename existing data structure AXObjectCache::m_nodeObjectMapping to AXObjectCache::m_nodeIdMapping, and make a
     new data structure with the name AXObjectCache::m_nodeObjectMapping that points from a Node& directly to an Ref<AccessibilityObject>.
     This removes one HashMap lookup in AXObjectCache::get. m_renderObjectMapping and m_widgetObjectMapping were renamed
     to m_renderObjectIdMapping and m_widgetIdMapping to stay consistent with m_nodeIdMapping. The cost of this new
     data structure is very small on most sites, and even in the most extreme case of html.spec.whatwg.org, only adds
     0.56mb extra memory usage. I believe this is worth the cost for what we get (results below).

  2. Inline the `get` part of `getOrCreate(Node&)`, and move the non-inlined part to new method `getOrCreateSlow`.
     Based on 60 seconds of using VoiceOver on html.spec.whatwg.org, we hit this new inlined fast path 98.9% of the
     time (~51,114,000 fast path hits vs. `590,000` slow-path hits). This required creation of a new method called
     exportedGetOrCreate which is outlined for consumers outside WebCore, as inlined functions cannot be exported.

  3. Add a new `AXObjectCache::getOrCreate(Element&)` method, allowing us to skip an `is<Document>` check in contexts
     where we've already casted a `Node` into an `Element`.

This eliminates 1532 samples (1.5 seconds) of main-thread time based on 60 seconds of using VoiceOver on html.spec.whatwg.org.

Places that used `getOrCreate` are now changed to import `AXObjectCacheInlines.h` instead of `AXObjectCache.h`.

* Source/WebCore/accessibility/AXListHelpers.cpp:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::cacheAndInitializeWrapper):
(WebCore::AXObjectCache::exportedGetOrCreate):
(WebCore::AXObjectCache::getOrCreateSlow):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::onPaint const):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXObjectCacheInlines.h:
(WebCore::AXObjectCache::getOrCreate):
* Source/WebCore/accessibility/AXSearchManager.cpp:
* Source/WebCore/accessibility/AccessibilityImageMapLink.cpp:
* Source/WebCore/accessibility/AccessibilityListBoxOption.cpp:
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
* Source/WebCore/accessibility/AccessibilityMenuListOption.cpp:
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::nextSibling const):
* Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
* Source/WebCore/inspector/InspectorOverlay.cpp:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/PrintContext.cpp:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::axObjectForElement const):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::getAccessibilityObjectForNode):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(-[WKPDFPluginAccessibilityObject parent]):
(-[WKPDFPluginAccessibilityObject accessibilityFocusedUIElement]):
(-[WKPDFPluginAccessibilityObject accessibilityAssociatedControlForAnnotation:]):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::accessibilityCoreObject):
* Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
(-[WKAccessibilityPDFDocumentObject accessibilityFocusedUIElement]):
(-[WKAccessibilityPDFDocumentObject accessibilityParent]):
(-[WKAccessibilityPDFDocumentObject accessibilityAssociatedControlForAnnotation:]):

Canonical link: https://commits.webkit.org/299439@main
